femtools/Particles.F90

```diff
@@ -525,33 +627,52 @@ subroutine read_particles_from_python(n_particles, subgroup_name, subgroup_path,
     type(attr_counts_type), intent(in) :: attr_counts
     !> Whether to consider this particle in a global element search
     logical, intent(in), optional :: global
-
-    integer :: i, str_size, proc_num
+    !> Number of particles being initialized
+    integer, intent(out) :: n_particles
+    !> ID number of last particle currently in list
+    integer, optional, intent(in) :: id_number
+    !> Python script used by initialise_during_simulation
+    character(len=PYTHON_FUNC_LEN), optional, intent(in) :: script
+
+    integer :: i, proc_num, stat
     character(len=PYTHON_FUNC_LEN) :: func
-    character(len=FIELD_NAME_LEN) :: fmt, particle_name
     real, allocatable, dimension(:,:) :: coords ! all particle coordinates, from python
+    ! if we don't know how many particles we're getting, we need a C pointer
+    type(c_ptr) :: coord_ptr
+    ! and a fortran pointer
+    real, pointer, dimension(:,:) :: coord_array_ptr
     real :: dt
 
     proc_num = getprocno()
 
     ewrite(2,*) "Reading particles from options"
-    call get_option(trim(subgroup_path)//"/initial_position/python", func)
+
+    if (present(script)) then
+       func=script
+    else
+       call get_option(trim(subgroup_path)//"/initial_position/python", func)
+    end if
     call get_option("/timestepping/timestep", dt)
+
+    call set_detectors_from_python(func, len(func), dim, current_time, coord_ptr, n_particles, stat)
+    call c_f_pointer(coord_ptr, coord_array_ptr, [dim, n_particles])
     allocate(coords(dim, n_particles))
-    call set_detector_coords_from_python(coords, n_particles, func, current_time)
+    if (n_particles==0) return
+    coords = coord_array_ptr
 
-    str_size = len_trim(int2str(n_particles))
-    fmt="(a,I"//int2str(str_size)//"."//int2str(str_size)//")"
+    call deallocate_c_array(coord_ptr)
```

Comment:
I guess this is okay after the `return` because the return is only for zero particles, in which case the pointer would be 0-length, so it's not going to leak memory...